text/0162-refactoring-support.md

```diff
@@ -0,0 +1,829 @@
+    Deployed?      No            Yes
+                          ("A" doesn't exist)
+
+**Scenario 2**: Due to skipped deployments, the refactor may end up referring to
+logical IDs that refer to different resources than intended. Suppose that, at
+some point, you had a resource called "A". In subsequent versions, "A" was
+deleted, and then replaced with a different resource that happened to also be
+named "A". If a mapping file references "A", the intention would be the most
+recent resource. But the resource that is found is the old one. In this case,
+the refactor will happen, but with a different result than intended:
+
+    Action:   [Create "A"]  [Delete "A"]  [Create diff. "A"]  [Rename "A" to "B"]
+                   │             │          │                     │
+    Time:      ────┼─────────────┼──────────┼─────────────────────┼─────────────>
+                   │             │          │                     │
+    Deployed?     Yes            No         No                   Yes
+                                                     (Applies to wrong resource)
+
+Keeping refactor files both in sync with the state of the CDK app, and also
+consistent with the states where it will be applied is a difficult problem. We
+recommend that you use refactor files sparingly, only in cases that automatic
+refactoring doesn't cover, such as ambiguity (see next section).
+
+#### Ambiguity
+
+You may find yourself in the very unlikely situation in which two or more
+resources have the same type and properties, and they are moved or renamed at
+the same time. Since there are at least two valid ways to map the old IDs to the
+new ones, the CLI can't guarantee that it will make the right choice, according
+to your intention.
+
+For instance, you may have two S3 buckets with identical sets of properties, and
+logical IDs `FinancialData` and `CustomerData`. You want to rename them to
+`FinancialReports` and `CustomerInfo`, respectively. But the CLI could end up
+doing the opposite, so that, after the refactor, the bucket that contains
+financial reports would have the logical ID `CustomerData`, and vice versa.
+
+Even though all the resources involved in this scenario would remain unchanged,
+we have decided to err on the side of caution and not perform the refactor. If,
+as the result of computing a refactor, the CLI detects such a case of ambiguity,
+it will fail with an error message explaining the situation.
+
+If you still want to perform a refactor, you have to explicitly provide the
+mapping. There are two ways to do this: you can either create a refactor file
+from scratch, or you can pass another command line option to only override the
+ambiguous part of the mapping.
+
+For example, if the CLI detects that there are two resources in `StackA`, `X`
+and `Y`, that are identical, and both were renamed to `Y` and `Z`, you can tell
+it exactly which mapping you want:
+
+    cdk refactor --override='{"StackA.X": "StackA.Z", "StackA.Y": "StackA.W"}'
```

Comment:
Can we make this slightly nicer to type for humans?

```suggestion
    cdk refactor --map StackA.X:StackA.Z --map StackA.Y:StackA.W
```

Reply:
I like this. Thanks!